tests/federation_connectivity_test.go

```diff
@@ -96,61 +98,61 @@ func TestNewUserCannotGetKeysForOfflineServer(t *testing.T) {
 // B will be able to decrypt C's message.
 // This is ultimately checking that Olm sessions are per-device and not per-room.
 func TestExistingSessionCannotGetKeysForOfflineServer(t *testing.T) {
-	// normally we would use ForEachClient or ClientTypeMatrix but due to federation we need
-	// to pin it to certain langs/servers.
-	tc := CreateTestContext(t, api.ClientType{
-		Lang: api.ClientTypeRust,
-		HS:   "hs1",
-	}, api.ClientType{
-		Lang: api.ClientTypeJS,
-		HS:   "hs2",
-	}, api.ClientType{
-		Lang: api.ClientTypeRust,
-		HS:   "hs1",
-	})
-	roomIDbc := tc.CreateNewEncryptedRoom(t, tc.Charlie, EncRoomOptions.Invite([]string{tc.Bob.UserID}))
-	roomIDab := tc.CreateNewEncryptedRoom(t, tc.Alice, EncRoomOptions.Invite([]string{tc.Bob.UserID}))
-	t.Logf("%s joining rooms %s and %s", tc.Bob.UserID, roomIDab, roomIDbc)
-	tc.Bob.MustJoinRoom(t, roomIDab, []string{"hs1"})
-	tc.Bob.MustJoinRoom(t, roomIDbc, []string{"hs1"})
-
-	tc.WithAliceBobAndCharlieSyncing(t, func(alice, bob, charlie api.Client) {
-		// let clients sync device keys
-		time.Sleep(time.Second)
-
-		// ensure encrypted messaging works in rooms ab,bc
-		wantMsgBody := "Hello world"
-		waiter := bob.WaitUntilEventInRoom(t, roomIDab, api.CheckEventHasBody(wantMsgBody))
-		evID := alice.SendMessage(t, roomIDab, wantMsgBody)
-		t.Logf("bob (%s) waiting for event %s", bob.Type(), evID)
-		waiter.Wait(t, 5*time.Second)
-		waiter = bob.WaitUntilEventInRoom(t, roomIDbc, api.CheckEventHasBody(wantMsgBody))
-		evID = charlie.SendMessage(t, roomIDbc, wantMsgBody)
-		t.Logf("bob (%s) waiting for event %s", bob.Type(), evID)
-		waiter.Wait(t, 5*time.Second)
-
-		// now bob's HS becomes unreachable
-		tc.Deployment.PauseServer(t, "hs2")
-
-		// C now joins the room ab
-		tc.Alice.MustInviteRoom(t, roomIDab, tc.Charlie.UserID)
-		tc.Charlie.MustJoinRoom(t, roomIDab, []string{"hs1"})
-
-		// let charlie sync device keys...
-		time.Sleep(time.Second)
-
-		// send a message as C: everyone should be able to decrypt this because Olm sessions
-		// are per-device, not per-room.
-		wantDecryptableMsgBody := "Bob can see this even though his server is down as we had a session already"
-		waiter = alice.WaitUntilEventInRoom(t, roomIDab, api.CheckEventHasBody(wantDecryptableMsgBody))
-		decryptableEventID := charlie.SendMessage(t, roomIDab, wantDecryptableMsgBody)
-		t.Logf("alice (%s) waiting for event %s", alice.Type(), decryptableEventID)
-		waiter.Wait(t, 5*time.Second)
-
-		// now bob's server comes back online
-		tc.Deployment.UnpauseServer(t, "hs2")
-
-		waiter = bob.WaitUntilEventInRoom(t, roomIDab, api.CheckEventHasBody(wantDecryptableMsgBody))
-		waiter.Wait(t, 10*time.Second) // longer time to allow for retries
+	ForEachClientType(t, func(t *testing.T, clientType api.ClientType) {
+		tc := CreateTestContext(t, api.ClientType{
+			Lang: clientType.Lang,
+			HS:   "hs1",
+		}, api.ClientType{
+			Lang: clientType.Lang,
```

Comment:
Same comment: this is now single-language and it used to be cross-language, right? Could we add a Lang2 that is populated with the other language where that is allowed?

Reply:
The test isn't actually designed to be testing different clients. It was doing that before because we couldn't use rust on hs2 so had to specify the language.